src/apps/polymer/IncentivizedPolymerEscrow.sol

```diff
@@ -0,0 +1,166 @@
+
+        // Send ack:
+        return AckPacket({success: true, data: receiveAck});
+    }
+
+    // The escrow manages acks, so any message can be directly provided to _handleAck.
+    function onUniversalAcknowledgement(bytes32 channelId, UniversalPacket calldata packet, AckPacket calldata ack)
+        external
+        onlyIbcMw
+    {
+        uint256 gasLimit = gasleft();
+        bytes32 feeRecipitent = bytes32(uint256(uint160(tx.origin)));
+
+        bytes calldata rawMessage = ack.data;
+        bytes memory destinationImplementationIdentifier = bytes.concat(packet.destPortAddr);
+
+        isVerifiedMessageHash[keccak256(rawMessage)] = VerifiedMessageHashContext({
+            chainIdentifier: channelId,
+            implementationIdentifier: destinationImplementationIdentifier
+        });
+        _handleAck(channelId, destinationImplementationIdentifier, rawMessage, feeRecipitent, gasLimit);
+    }
+
+    // For timeouts, we need to construct the message.
+    function onTimeoutUniversalPacket(bytes32 channelId, UniversalPacket calldata packet) external onlyIbcMw {
+        uint256 gasLimit = gasleft();
+        bytes32 feeRecipitent = bytes32(uint256(uint160(tx.origin)));
+
+        bytes calldata rawMessage = packet.appData;
+        bytes32 messageIdentifier = bytes32(rawMessage[MESSAGE_IDENTIFIER_START:MESSAGE_IDENTIFIER_END]);
+        address fromApplication = address(uint160(bytes20(rawMessage[FROM_APPLICATION_START_EVM:FROM_APPLICATION_END])));
+        _handleTimeout(
+            channelId, messageIdentifier, fromApplication, rawMessage[CTX0_MESSAGE_START:], feeRecipitent, gasLimit
+        );
+    }
+
+    /**
+     * @param destinationChainIdentifier  Universal Channel ID. It's always from the running chain's perspective.
+     * Each universal channel/channelId represents a directional path from the running chain to a destination chain.
+     * Universal ChannelIds should _destChainIdToChannelIdd from the Polymer registry.
+     * Although everyone is free to establish their own channels, they're not "officially" vetted until they're in the Polymer registry.
+     * @param destinationImplementation IncentivizedPolymerEscrow address on the counterparty chain.
+     * @param message packet payload
+     * @param deadline Packet will timeout after the dest chain's block time in nanoseconds since the epoch passes timeoutTimestamp.
+     */
+    function _sendPacket(
+        bytes32 destinationChainIdentifier,
+        bytes memory destinationImplementation,
+        bytes memory message,
+        uint64 deadline
+    ) internal override returns (uint128 costOfsendPacketInNativeToken) {
+        // If timeoutTimestamp is set to 0, set it to maximum. This does not really apply to Polymer since it is an onRecv implementation
+        // but it should still conform to the general spec of Generalised Incentives.
+        uint64 timeoutTimestamp = deadline > 0 ? deadline : type(uint64).max;
```

Comment:
@alfredo-stonk Is this allowed or should I be more conservative in the way deadline?

When I looked at your contracts, it looked like the channel is ordered. If I do this and the message doesn't get relayed is there anything bad?

This also applies to general unbounded deadlines.

We are already checking if deadline is in the past, so that isn't a problem. We do make an exception when the deadline is 0, in those cases we want the message to not be timed out. Is that a problem?

If yes, what should the maximum deadline be?

Reply:
> Is this allowed or should I be more conservative in the way deadline?

This is allowed

> When I looked at your contracts, it looked like the channel is ordered. If I do this and the message doesn't get relayed is there anything bad?

Our IBC Dispatcher supports both ordered and unordered channel types. But UniversalChannel IBC Middleware contract only works with _unordered_  channels for now, so that undelivered packets won't clog later packets.


> We are already checking if deadline is in the past, so that isn't a problem. We do make an exception when the deadline is 0, in those cases we want the message to not be timed out. Is that a problem?
If yes, what should the maximum deadline be?

Setting unbounded packet timeout to max uint64 is totally ok